chapters/grid-search.qmd

```diff
@@ -0,0 +1,390 @@
+
+
+::: {#alg-grid-search}
+
+:::: {.columns}
+
+::: {.column width="10%"}
+
+:::
+
+::: {.column width="80%"}
+
+
+```pseudocode
+#| html-line-number: true
+#| html-line-number-punc: ":"
+
+\begin{algorithm}
+\begin{algorithmic}
+\State $\mathfrak{D}^{tr}$: training set of predictors $X$ and outcome $y$
+\State $B$: number of resamples
+\State $M(\mathfrak{D}^{tr}, B)$: a mapping function to split $\mathfrak{D}^{tr}$ for each of  $B$ iterations.
+\State $f()$: model pipeline 
+\State $\Theta$: Parameter set ($s \times m$) with candidates $\theta_j$
+\For{$j=1$ \To $s$}
+  \State Generate $\hat{Q}_{j} =$ \Call{Resample}{$\mathfrak{D}^{tr}, f(\cdot;\theta_j), M(\mathfrak{D}^{tr}, B)$} corresponding to candidate $\theta_j$.
+\EndFor
+\State Determine $\hat{\theta}_{opt}$ that optimizes $\hat{Q}_{j}$.
+\end{algorithmic}
+\end{algorithm}
+```
+
+:::
+
+::: {.column width="10%"}
+
+:::
+
+::::
+
+Grid search for tuning parameter optimization that loops over the resampling algorithm shown in @alg-resample.
+
+:::
+
+
+```{r}
+#| label: grids
+#| include: false
+
+bst_spec <- 
+  boost_tree(learn_rate = tune(), trees = tune()) %>% 
+  set_mode("classification")%>% 
+  set_engine("lightgbm")
+
+bst_wflow <-
+  workflow() %>%
+  add_model(bst_spec) %>%
+  add_formula(class ~ A + B)
+
+bst_param <- 
+  bst_wflow %>% 
+  extract_parameter_set_dials() %>% 
+  update(learn_rate = learn_rate(c(-3, -1)))
+
+bst_reg <-
+  grid_regular(bst_param, levels = c(5, 3)) %>%
+  mutate(grid = "Regular")
+
+sfd_size <- nrow(bst_reg)
+
+set.seed(nrow(bst_reg))
+bst_random <-
+  grid_random(bst_param, size = sfd_size) %>%
+  mutate(grid = "Random")
+
+bst_sfd <- grid_space_filling(bst_param, size = sfd_size) %>%
+  mutate(grid = "Space-Filling")
+
+seq_tree <- sort(unique(bst_reg$trees))
+txt_tree <- knitr::combine_words(seq_tree)
+
+seq_ord <- format(sort(unique(log10(bst_reg$learn_rate))))
+seq_ord <- paste0("10<sup>", seq_ord, "</sup>")
+lr_ord <- knitr::combine_words(seq_ord)
+
+bst_grids <- 
+  bind_rows(bst_reg, bst_random, bst_sfd) %>%
+  mutate(
+    grid = factor(grid, levels = c("Regular", "Random", "Space-Filling"))
+  )
+```
+
+To demonstrate, this chapter will initially focus on grids for a boosted tree model with two tuning parameters. Recall that a boosted tree is a collection of individual decision trees created sequentially. One parameter related to the process of creating the ensemble is the learning rate: the next tree uses information from the last tree to improve it. An important parameter is how much, or how _fast_ it learns. It could be that some data sets desire a large number of trees that evolve slowly (i.e., low learning rate), and other data sets work best with fewer trees that change rapidly. The learning rate parameter must be greater than zero. Boosted tree models typically work in the range of 10<sup>-5</sup> to  10<sup>-1</sup>. This parameter is best conceptualized in log units. 
```

Comment:
```suggestion
To demonstrate, this chapter will initially focus on grids for a boosted tree model with two tuning parameters. Recall that a boosted tree is a collection of individual decision trees created sequentially. One parameter related to the process of creating the ensemble is the learning rate: the next tree uses information from the last tree to improve it. An important parameter is how much, or how _fast_ it learns. It could be that some data sets need a large number of trees that evolve slowly (i.e., low learning rate) to find an optimal value of the performance metric.  Alternatively, other data sets require fewer trees that change rapidly for optimal predictive performance. The learning rate parameter must be greater than zero, and is typically in the range of 10<sup>-5</sup> to  10<sup>-1</sup>.  Because the magnitude of this parameter is small, it is best conceptualized in log units. 
```